### ISSUE TITLE
Running node.js script in Android Java app using nodejs-mobile

### ISSUE THREAD

**Issue by adonnini** [janeasystems/nodejs-mobile#66]
For reference, what follows is the initial issue entry posted on nodejs/help, and responses.

At this point, I am following @orangemocha's  advice and looking into using nodejs-mobile.

First question:

Is there a reason why I would not just follow the instructions in 

https://code.janeasystems.com/nodejs-mobile/getting-started-android

instead of using the react-native plugin?

Thanks,

Alex Donnini

================================================
RESPONSE #5 

@adonnini running complex JS scripts is no problem. The part that might not be easy out of the box is marshalling data between Java and JavaScript. But it's doable, and there is code that does it in the react-native and cordova plug-ins, that you could copy.
Anyway, nodejs-mobile might be a better place of this discussion smiley

===================================================
RESPONSE #4 (me)

Thanks for suggestion. As they say, been there, done that. I could be wrong (would love to be wrong). It does not work for Android Java applications that need to access complex JS scripts. Thanks anyway, Alex Donnini

=================================================
RESPONSE #3 

You might want to try https://github.com/janeasystems/nodejs-mobile

==================================================
RESPONSE #2 (me)
Thanks. I am familair with the blog. Unfortunately, it does not solve the problem I have.

When trying to run node.js via J2V8, my app crashes with the following error:

java.lang.UnsupportedOperationException: StartNodeJS Not Supported.

When I brought this to the attention of the folks at EclipseSource, the response was:

"The node wrappers (for running node.js in an android app) are not available on Android, they are only available on the Desktop platforms (windows, mac, linux). This is the expected behaviour until we have node binaries for Android."

My current concern is that J2V8 is basically dormant. It would be a real shame.

I am looking for alternatives. That's what brought me here.

Can you help in any way?

Thanks,

Alex Donnini

==========================================
RESPONSE #1
    "I tried using J2V8. It does not have Android wrappers for running Node.js"

https://eclipsesource.com/blogs/2016/07/20/running-node-js-on-the-jvm/

hope i helped

=============================================
INITIAL POST

Hi,

Apologies, if this is not the place to ask this question. If it is not, please let me know where I should ask it.

I need to run a JavaScript in my Android Java app (a blockchain driver). There is no UI. No need/desire to run WebView.

I tried using J2V8. It does not have Android wrappers for running Node.js

Any help/ideas on how to do this would be greatly appreciated.

Thanks,

Alex Donnini

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,
Thank you for the question.

The base Android release of "nodejs-mobile" is a direct shared library build of node, so you'll have more flexibility with what you can do with it but you'll need to build what you need on top of it (as an example, you need to develop a communication channel between your Application's activity and the background node server). The getting started example uses a web server in the node side to get requests from the Activity. Samples for Android are also pushed to this repo: https://github.com/janeasystems/nodejs-mobile-samples/tree/master/android

The "nodejs-mobile-cordova" and "nodesjs-mobile-react-native" have more functionality built on the plugin side: an in-process communication channel to emit events between the UI JavaScript runtime and the nodejs-mobile JavaScript runtime, copy of a nodejs-project folder from the assets to the Application's data folder so node can have the filesystem structure it expects to `require` modules properly and an experimental native modules support you need to manually turn on currently.

To sum it up, the Android library gives you more flexibility (and work) in how you integrate it with your Application, while the plugins are better suited to getting started fast using either of the two UI JavaScript frameworks for mobile.

I hope this is helpful.

**Comment by adonnini** [comment]
@jaimecbernardo thank you this is very helpful.

I will look into the react-native plugin.

A little more detail regarding what I am trying to do to make sure I am on the right track. 

In order to process a blockchain transaction my app needs to use the JS bigchaindb driver (https://github.com/bigchaindb/js-bigchaindb-driver). Bigchaindb provides a package for use with a browser or other UI, not what I need.

I need to be able to access the driver from an Android service. I downloaded the driver source. Using Eclipse-JS I built a node using the Eclipse npm build function.

Now, I have a node folder containing the bigchaindb js driver files.

This is my starting point for trying to use your software to access the bigchaindb js driver. 

Am I on the right track?

Thanks

**Comment by jaimecbernardo** [comment]
I'm not familiar with the JS module mentioned or what Eclipse-JS does, but if eclipse-JS is building native modules for you, it's probably building them for the system Eclipse is running on, so they might not run on Android. Instructions on how to turn on the native modules experimental feature in case you have that issue can be found on this repo's issue regarding native modules: https://github.com/janeasystems/nodejs-mobile/issues/4#issuecomment-371555631
The best approach is to give it a try.

Please let us know how it goes.

**Comment by adonnini** [comment]
Thanks. I will take a look at issue #4 

Before I put in a lot of time and effort into this approach I want to make sure I understand what you think might work. Sorry for being pedantic.

You are suggesting that if I take the JS bigchain driver, and package it as a react application, then I can use the react-native plugin to use/access the JS bigchaindb driver from my Android Java app. 

Is this what you are suggesting might work?

Thanks for your patience

Alex

**Comment by adonnini** [comment]
Hi again,

I just read through this very good posting
https://medium.com/mindorks/react-native-integration-with-existing-app-f2757c2e672d

I am afraid that react is not the right vehicle for accomplishing what I am trying to do.

React is clearly UI and browser oriented. I am looking to use a driver in a service with no UI.

If I were using a java driver, I would simply be able to use the driver's functions (method and class calls) in my app by including the driver in my project's dependencies (or adding it to my apk). 

I am looking to do the same with a driver which is not written in Java and packaged in a jar but is written in JS. I am not partial to one form of packaging vs. another.

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,
The native Android library would be the way to go for your case, then, by also developing a way to communicate with the Node.js engine so you could call those functions.

**Comment by adonnini** [comment]
hi @jaimecbernardo,

Quick question. Can startNodeWithArguments call a multi-module node.js script stored on sdcard? From the looks of it, it can't (Obviously I could be completely wrong). 

How would/could I call a multi-module node.js script?

Thanks,

Alex

**Comment by adonnini** [comment]
I think I answered my own question. I found the source of MainActivity.java in 

nodejs-mobile-samples-master/android/native-gradle-node-folder/app/src/main/java/com/yourorg/sample/

with the following:

            new Thread(new Runnable() {
                @Override
                public void run() {
                    //The path where we expect the node project to be at runtime.
                    String nodeDir=getApplicationContext().getFilesDir().getAbsolutePath()+"/nodejs-project";
                    if (wasAPKUpdated()) {
                        //Recursively delete any existing nodejs-project.
                        File nodeDirReference=new File(nodeDir);
                        if (nodeDirReference.exists()) {
                            deleteFolderRecursively(new File(nodeDir));
                        }
                        //Copy the node project from assets into the application's data path.
                        copyAssetFolder(getApplicationContext().getAssets(), "nodejs-project", nodeDir);

                        saveLastUpdateTime();
                    }
                    startNodeWithArguments(new String[]{"node",
                            nodeDir+"/main.js"
                    });
                }
            }).start();

Thanks,

Alex

**Comment by adonnini** [comment]
Good morning. I followed the instructions in 
https://code.janeasystems.com/nodejs-mobile/getting-started-android
and implemented the sample app.

When I rant it on my phone, it failed producing  the following log:

04-29 10:17:17.880 9191-9191/? I/art: Late-enabling -Xcheck:jni
04-29 10:17:18.035 9191-9191/commmanager.android.com.nodejstest I/InstantRun: starting instant run server: is main process
04-29 10:17:18.151 9191-9191/commmanager.android.com.nodejstest W/art: Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
04-29 10:17:18.247 9191-9208/commmanager.android.com.nodejstest E/art: No implementation found for java.lang.Integer commmanager.android.com.nodejstest.MainActivity.startNodeWithArguments(java.lang.String[]) (tried Java_commmanager_android_com_nodejstest_MainActivity_startNodeWithArguments and Java_commmanager_android_com_nodejstest_MainActivity_startNodeWithArguments___3Ljava_lang_String_2)
04-29 10:17:18.248 9191-9208/commmanager.android.com.nodejstest E/AndroidRuntime: FATAL EXCEPTION: Thread-4
    Process: commmanager.android.com.nodejstest, PID: 9191
    java.lang.UnsatisfiedLinkError: No implementation found for java.lang.Integer commmanager.android.com.nodejstest.MainActivity.startNodeWithArguments(java.lang.String[]) (tried Java_commmanager_android_com_nodejstest_MainActivity_startNodeWithArguments and Java_commmanager_android_com_nodejstest_MainActivity_startNodeWithArguments___3Ljava_lang_String_2)
        at commmanager.android.com.nodejstest.MainActivity.startNodeWithArguments(Native Method)
        at commmanager.android.com.nodejstest.MainActivity$1.run(MainActivity.java:41)
        at java.lang.Thread.run(Thread.java:761)
04-29 10:17:18.435 9191-9191/commmanager.android.com.nodejstest W/art: Before Android 4.1, method int android.support.v7.widget.ListViewCompat.lookForSelectablePosition(int, boolean) would have incorrectly overridden the package-private method in android.widget.ListView

Where did I go wrong?

Thanks,

Alex

**Comment by adonnini** [comment]
my MainActivity.java is:

package commmanager.android.com.nodejstest;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.TextView;

import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.os.AsyncTask;
import java.net.*;
import java.io.*;


public class MainActivity extends AppCompatActivity {

    static {
        System.loadLibrary("native-lib");
        System.loadLibrary("node");
    }

    // Used to load the 'native-lib' library on application startup.
    //static {
    //    System.loadLibrary("native-lib");
    //}


    //We just want one instance of node running in the background.
    public static boolean _startedNodeAlready=false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if( !_startedNodeAlready ) {
            _startedNodeAlready=true;
            new Thread(new Runnable() {
                @Override
                public void run() {
                    startNodeWithArguments(new String[]{"node", "-e",
                            "var http = require('http'); " +
                                    "var versions_server = http.createServer( (request, response) => { " +
                                    "  response.end('Versions: ' + JSON.stringify(process.versions)); " +
                                    "}); " +
                                    "versions_server.listen(3000);"
                    });
                }
            }).start();
        }

        final Button buttonVersions = (Button) findViewById(R.id.btVersions);
        final TextView textViewVersions = (TextView) findViewById(R.id.tvVersions);

        buttonVersions.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {

                //Network operations should be done in the background.
                new AsyncTask<Void,Void,String>() {
                    @Override
                    protected String doInBackground(Void... params) {
                        String nodeResponse="";
                        try {
                            URL localNodeServer = new URL("http://localhost:3000/");
                            BufferedReader in = new BufferedReader(
                                    new InputStreamReader(localNodeServer.openStream()));
                            String inputLine;
                            while ((inputLine = in.readLine()) != null)
                                nodeResponse=nodeResponse+inputLine;
                            in.close();
                        } catch (Exception ex) {
                            nodeResponse=ex.toString();
                        }
                        return nodeResponse;
                    }
                    @Override
                    protected void onPostExecute(String result) {
                        textViewVersions.setText(result);
                    }
                }.execute();

            }
        });



        // Example of a call to a native method
    //    TextView tv = (TextView) findViewById(R.id.sample_text);
    //    tv.setText(stringFromJNI());
    }

    /**
     * A native method that is implemented by the 'native-lib' native library,
     * which is packaged with this application.
     */
    public native Integer startNodeWithArguments(String[] arguments);
//    public native String stringFromJNI();
}

**Comment by adonnini** [comment]
Problem solved. All set.

Next I will try and run the version accessing a script from file

**Comment by adonnini** [comment]
Update.

I have both native Android sample apps running. Before I use them as "template" for my app, I have two questions:

1) Does "Currently, only a single instance of the Node.js runtime can be started within an application. Restarting the engine after it has finished running is also not supported."

mean that I can run startNodeWithArguments only once? If my understanding is incorrect, could you give me an example?

2) Does "Having a nodejs-project path with a package.json inside is helpful for using npm modules, by running npm install {module_name} inside nodejs-project so that the modules are also packaged with the application and made available at runtime."

mean that if I wanted modules to be packaged with my app, I would need to run npm install? I am asking because I did and the "native-gradle-node-folder" app modified to use the bigchaindb-js driver, failed to build.

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,
1. You are correct. You should start it only once. The node runtime expects to be run once per process, so it's hard to evaluate if it's able to restart from a "clean slate" in the same process. It might be possible, but it's an untested scenario.
2. Runnin `npm install {module}` inside `nodejs-project` will create the `nodejs-project/node_modules/` path with the installed modules inside, so that they are copied to your Application's assets. Android's packaging default behavior can cause issues while trying to include some files. Do you get any error message while building that could give a int to what is causing the error?

**Comment by adonnini** [comment]
Thanks @jaimecbernardo Here is another update:

1) I have two startNodeWithArguments statements in the thread in MainActivity.java. Everything seems to be working normally

2) I resolved the problem regarding the node_modules conflict. Previously, I had run a build via Eclipse. This time, I ran npm install directly in the folder where the package.json for the script is located.

Next I need to run the bigchaindb connection and transaction related steps. I am not too optimistic that I'll be able to make it work. We'll see...

Thanks,

Alex

**Comment by adonnini** [comment]
Third update:

I started running scripts using the bigchaindb-js driver. The app failed with the following:

04-30 18:15:24.749: E/NODEJS-MOBILE(18759): fs.js:653
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):   return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):                  ^
04-30 18:15:24.749: E/NODEJS-MOBILE(18759): Error: EACCES: permission denied, open '/data/.babel.json'
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at Object.fs.openSync (fs.js:653:18)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at Object.fs.writeFileSync (fs.js:1300:33)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at save (/data/data/com.yourorg.sample/files/bigchaindb-js/node_modules/babel-register/lib/cache.js:48:16)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at _combinedTickCallback (internal/process/next_tick.js:131:7)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at process._tickCallback (internal/process/next_tick.js:180:9)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at Function.Module.runMain (module.js:667:11)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at startup (bootstrap_node.js:188:16)
04-30 18:15:24.749: E/NODEJS-MOBILE(18759):     at bootstrap_node.js:611:3

My guess is that because the sample app does not have permission to write to external storage. What do you think?

Thanks,

Alex

**Comment by adonnini** [comment]
Update on crash:

I added permissions to write to external storage. Assuming I did it correctly (the same code works in my location tracking app), it made no difference.

Thanks,

Alex

**Comment by adonnini** [comment]
2nd update on crash.

It looks like the problem is caused by babel attempting to create/write to a cache file in the android /data directory where it does not have permission.

I have been looking for a way to programatically set BABEL_DISABLE_CACHE to 1 and/or change the babel cache location to the sdcard. 

So far, I have been unable to do either. Help on this would be greatly appreciated.

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,
Indeed, writing in `/data/` seems to be too close to the system root for an Android application to have write permissions (this would be a big security concern if it was able too).
This is a module behavior you'd need to figure out how to do change. If it's the matter of setting environment variables, you can check this example on how we're setting the TMPDIR variable (Android doesn't really have a temp dir) in a cordova staging branch:
https://github.com/janeasystems/nodejs-mobile-cordova/blob/dfa18b4743f5154543939e7498a553a8429a925d/src/android/java/com/janeasystems/cdvnodejsmobile/NodeJS.java#L84

**Comment by adonnini** [comment]
Thanks. I have worked around the issue.

Next, the app crashes with the following:

05-01 18:40:46.014: E/filemap(20738): mmap(15536128,0) failed: Invalid argument
05-01 18:40:46.014: W/asset(20738): create map from entry failed
05-01 18:40:46.014: W/System.err(20738): java.io.FileNotFoundException: bigchaindb-js/node_modules/eslint-plugin-import/node_modules/resolve/test/resolver/other_path/root.js
05-01 18:40:46.014: W/System.err(20738): 	at android.content.res.AssetManager.openAsset(Native Method)
05-01 18:40:46.014: W/System.err(20738): 	at android.content.res.AssetManager.open(AssetManager.java:347)
05-01 18:40:46.014: W/System.err(20738): 	at android.content.res.AssetManager.open(AssetManager.java:321)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAsset(MainActivity.java:228)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:211)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.copyAssetFolder(MainActivity.java:215)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity.access$200(MainActivity.java:30)
05-01 18:40:46.014: W/System.err(20738): 	at com.yourorg.sample.MainActivity$1.run(MainActivity.java:86)
05-01 18:40:46.014: W/System.err(20738): 	at java.lang.Thread.run(Thread.java:761)

Based on some searches I have run, it looks like it may have t do with dependencies. Have you seen this kind of failure before? Any suggestions as to how I could resolve it?

Thanks,

Alex

**Comment by adonnini** [comment]
Yesterday's problem solved. At least, I worked around it.

Today's crash

05-02 21:58:26.168: W/Thread-4(9112): type=1400 audit(0.0:170): avc: denied { ioctl } for path="/sys/kernel/debug/tracing/trace_marker" dev="debugfs" ino=5954 ioctlcmd=5451 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:debugfs_trace_marker:s0 tclass=file permissive=0
05-02 21:58:26.168: W/Thread-4(9112): type=1400 audit(0.0:171): avc: denied { ioctl } for path="/dev/pmsg0" dev="tmpfs" ino=11593 ioctlcmd=5451 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:pmsg_device:s0 tclass=chr_file permissive=0
05-02 21:58:30.331: W/Thread-4(9112): type=1400 audit(0.0:172): avc: denied { ioctl } for path="/sys/kernel/debug/tracing/trace_marker" dev="debugfs" ino=5954 ioctlcmd=5451 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:debugfs_trace_marker:s0 tclass=file permissive=0
05-02 21:58:30.331: W/Thread-4(9112): type=1400 audit(0.0:173): avc: denied { ioctl } for path="/dev/pmsg0" dev="tmpfs" ino=11593 ioctlcmd=5451 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:pmsg_device:s0 tclass=chr_file permissive=0
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): #
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): # Fatal error in ../deps/v8/src/v8.cc, line 95
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): # 
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): Check failed: !platform_.
05-02 21:58:30.336: E/NODEJS-MOBILE(9059): #
05-02 21:58:30.336: A/libc(9059): Fatal signal 5 (SIGTRAP), code 1 in tid 9112 (Thread-4)
05-02 21:58:30.336: W/(384): debuggerd: handling request: pid=9059 uid=10105 gid=10105 tid=9112
05-02 21:58:30.487: A/DEBUG(9288): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
05-02 21:58:30.487: A/DEBUG(9288): Build fingerprint: 'google/angler/angler:7.1.2/N2G47H/3783593:user/release-keys'
05-02 21:58:30.487: A/DEBUG(9288): Revision: '0'
05-02 21:58:30.487: A/DEBUG(9288): ABI: 'arm64'
05-02 21:58:30.487: A/DEBUG(9288): pid: 9059, tid: 9112, name: Thread-4  >>> com.yourorg.sample <<<
05-02 21:58:30.487: A/DEBUG(9288): signal 5 (SIGTRAP), code 1 (TRAP_BRKPT), fault addr 0x7d5cff5958
05-02 21:58:30.487: A/DEBUG(9288):     x0   0000000000000000  x1   0000000000000000  x2   0000000000000000  x3   ffffffffffffffff
05-02 21:58:30.487: A/DEBUG(9288):     x4   0000000000000058  x5   0000000000000000  x6   0000007d5d7dddba  x7   6573606b6f20fefe
05-02 21:58:30.487: A/DEBUG(9288):     x8   0000000000000001  x9   0000000000000001  x10  0000000000004001  x11  0000000000000000
05-02 21:58:30.487: A/DEBUG(9288):     x12  0000000000002398  x13  000000000ccccccc  x14  000000008000002f  x15  0000007d780ff268
05-02 21:58:30.487: A/DEBUG(9288):     x16  0000007d5d40c228  x17  0000007d5cff5940  x18  0000000000000000  x19  0000007d7811c7c8
05-02 21:58:30.487: A/DEBUG(9288):     x20  0000007d5d2a11bc  x21  000000000000005f  x22  0000007d5d3356dd  x23  0000007d759d0ee8
05-02 21:58:30.487: A/DEBUG(9288):     x24  0000007d5d438d80  x25  a555a02dfa7ea075  x26  0000007d5d898498  x27  a555a02dfa7ea075
05-02 21:58:30.487: A/DEBUG(9288):     x28  0000007d5d7de6d0  x29  0000007d5d7de530  x30  0000007d5cff47bc
05-02 21:58:30.487: A/DEBUG(9288):     sp   0000007d5d7de410  pc   0000007d5cff5958  pstate 0000000080000000
05-02 21:58:30.493: A/DEBUG(9288): backtrace:
05-02 21:58:30.493: A/DEBUG(9288):     #00 pc 0000000000f33958  /data/app/com.yourorg.sample-2/lib/arm64/libnode.so (_ZN2v84base2OS5AbortEv+24)
05-02 21:58:30.493: A/DEBUG(9288):     #01 pc 0000000000f327b8  /data/app/com.yourorg.sample-2/lib/arm64/libnode.so (V8_Fatal+216)
05-02 21:58:31.553: E/(9288): debuggerd: failed to kill process 9059: No such process
05-02 21:58:31.556: W/ActivityManager(923):   Force finishing activity com.yourorg.sample/.MainActivity
05-02 21:58:31.563: W/ActivityManager(923):   Force finishing activity com.yourorg.sample/.MainActivity
05-02 21:58:31.563: W/ActivityManager(923): Duplicate finish request for ActivityRecord{eac1ec7 u0 com.yourorg.sample/.MainActivity t128 f}
05-02 21:58:31.587: W/(384): debuggerd: resuming target 9059
05-02 21:58:31.588: I/BootReceiver(923): Copying /data/tombstones/tombstone_01 to DropBox (SYSTEM_TOMBSTONE)
05-02 21:58:31.616: D/GraphicsStats(923): Buffer count: 6
05-02 21:58:31.617: I/ActivityManager(923): Process com.yourorg.sample (pid 9059) has died
05-02 21:58:31.617: D/ActivityManager(923): cleanUpApplicationRecord -- 9059
05-02 21:58:31.618: W/InputDispatcher(923): channel '5590908 com.yourorg.sample/com.yourorg.sample.MainActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
05-02 21:58:31.618: E/InputDispatcher(923): channel '5590908 com.yourorg.sample/com.yourorg.sample.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
05-02 21:58:31.619: I/Zygote(549): Process 9059 exited due to signal (5)
05-02 21:58:31.621: I/WindowManager(923): WIN DEATH: Window{5590908 u0 com.yourorg.sample/com.yourorg.sample.MainActivity}
05-02 21:58:31.621: W/InputDispatcher(923): Attempted to unregister already unregistered input channel '5590908 com.yourorg.sample/com.yourorg.sample.MainActivity (server)'

Frankly, diagnosing this latest problem is beyond my skills.

Can you help?

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,
From the crash report:
```
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): #
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): # Fatal error in ../deps/v8/src/v8.cc, line 95
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): #
05-02 21:58:30.335: E/NODEJS-MOBILE(9059): Check failed: !platform_.
05-02 21:58:30.336: E/NODEJS-MOBILE(9059): #
```
Looking at the referred source code: https://github.com/janeasystems/nodejs-mobile/blob/f745cdb90b4c8dea8bcc7d8f36c7706a29e2af37/deps/v8/src/v8.cc#L94-L100

I suspect it is something to do with trying to start the engine twice, as it seems like a check in the V8 Javascript Engine to fail in Initializing the Platform if it has not been shutdown before.

**Comment by adonnini** [comment]
I guessed as much. 

How big a deal is to fix this bug? After all, J2V8, for example, lets users start and stop the engine any number of times, and lets users run more than on instance of the engine at the same time.

I think that in a production environment this is a pretty significant limitation.

It makes it impossible to load/run a node.js to run a script in an Android app.

Thanks,

Alex

**Comment by adonnini** [comment]
By the way, I guess that setting up multiple threads to run instances of the engine would not make any difference, right?

**Comment by jaimecbernardo** [comment]
Hi @adonnini,

I'm not aware of how J2V8 works, but I think that to let users run more than one instance of the engine, they'd have to change it, start each instance on a different process, or initialize the engine or event loop differently. In case J2V8 initializes the engine differently, the same APIs would be present in the "nodejs-mobile" shared library, like `node::Init`, so you could call them instead of `node::Start`. The code "nodejs-mobile" code base tries to be as similar as possible with the base project, so another technique for embedding node that you find on another project should also work with "nodejs-mobile".

The current scenario "nodejs-mobile" supports is by embedding the "nodejs-mobile" runtime in the application's process, and this is supported for both Android and iOS, which is a big goal of the project. A way to get around this in Android would be to build node as an executable that you embed in your application and start it in a new process as if it was being started from the command line, but this wouldn't be supported on iOS.

Regarding the running of different scripts in an Android app, the use case can be supported by making a main script that runs in the background and sets a communication layer that enables your app to control and call different code depending on what you send through the communication channel.

> By the way, I guess that setting up multiple threads to run instances of the engine would not make any difference, right?

Multiple threads would not make any difference, as the engine is shared in the process. Multiple processes should you enable you to do this, however.

Please let us know of the results you get if you try any of the methods.

**Comment by adonnini** [comment]
Hi @jaimecbernardo 

Thanks for the feedback. 

When you say 

"the same APIs would be present in the "nodejs-mobile" shared library, like node::Init, so you could call them instead of node::Start."
the API you mention is not accessible via an Android Java app, is it? How?

After writing the message above, I realized that, as you say, multiple processes would resoolve the issue except for the fact that the scripts running in the different processes would need to communicate with each other. I am not sure this would be doable with the current implementation.

Although "build node as an executable that you embed in your application and start it in a new process as if it was being started from the command line" would probably solve the problem, I don't think it would be a good way to go. I agree.

this option

"Regarding the running of different scripts in an Android app, the use case can be supported by making a main script that runs in the background and sets a communication layer that enables your app to control and call different code depending on what you send through the communication channel."

is the one that might make the most sense for me to pursue (pending a clarification regarding access to the API (above)).

If I understand you correctly, I would need a script that when run from my app fires off the JS bigchaindb driver in the background, and executes tasks as needed.

However, even under this scenario I would still not be able to run more than one startNodeWithArguments per process, correct?

If the app tries to submit more than one task to the main script, how do yopu envisage communication of these task taking place if not via multiple startNodeWithArguments executions?

Sorry if I am not being clear, or misunderstood what you said in your message. Before proceeding, I wanted to be as clear as possible regarding what you are thinking.

Thanks,

Alex

**Comment by adonnini** [comment]
Sorry, I forgot to ask you. When you mention "communication layer" what did you have in mind? 

Frankly, I am still struggling to see how this approach (background script + comm layer) would not require multiple executions of startNodeWithArguments. Would I use another call to run code that accesses the background script via the comm layer. By the way, this approach is the one I like the most.

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi, @adonnini ,

> "the same APIs would be present in the "nodejs-mobile" shared library, like node::Init, so you could call them instead of node::Start."
the API you mention is not accessible via an Android Java app, is it? How?

You'd have to wrap it around JNI, like `node::Start()` in the sample: https://github.com/janeasystems/nodejs-mobile-samples/tree/64b556e0f8b67a165cd5012e17156087f39d8f73/android/native-gradle#add-native-jni-function-to-start-nodejs

> However, even under this scenario I would still not be able to run more than one startNodeWithArguments per process, correct?

Correct.

> If the app tries to submit more than one task to the main script, how do yopu envisage communication of these task taking place if not via multiple startNodeWithArguments executions?

There could be many ways of doing this:
- It could be a web server running in nodejs, like in the sample, but handling different requests.
- It could be through "pipes" set on both sides.
- It could be through a message channel set in native code like it's done in the cordova and react-native plugins.

Perhaps there is some detail I'm also missing, due to not being familiar to what you're trying to do, since using `startNodeWithArguments` many times seems to be important to you. Would you be able to provide more details in what you're trying to do and why this would be the case?

**Comment by adonnini** [comment]
Hi @jaimecbernardo 

I want to create, execute and track a bigchaindb blockchain transaction from my app which is not and cannot be browser based.

In order to do that I need to use the bigchaindb-js driver (https://github.com/bigchaindb/js-bigchaindb-driver).

Currently, the bigchiandb-js driver is not packaged for use in an Android Java app which is not browser based.

I downloaded the source for the bigchaindb-js driver and ran npm install on it.

I stored the resulting set of folders (all of them) in app/src/main/assets/bigchaindb-js in native-gradle-node-folder sample Android java app taken from your github site.

I successfully built the native-gradle-node-folder app and ran it.

MainActivity.java has two  startNodeWithArguments statements:

1) The first to run the bighchaindb-js driver

startNodeWithArguments(new String[]{"node", nodeDir + "/src/index.js"

2) The second to create and execute a bigchaindb blockchain transaction

Based on the application log, the bigchaindb-js driver does seem to load an run correctly.

However, as you know, the application crashes because of the second startNodeWithArguments statement which triggers an attempt to run a second node.js runtime environment.

I think the problem I have may be due to my trying to use a model of execution that works in an all-Java application, i.e. have the bigchaindb driver as a dependency and then create and execute a bigchaindb blockchain transaction using methods in the driver.

I don't think having a single startNodeWithArguments statement for the creation and execution of a bigchaindb blockchain transaction qith "require" (?) or equivalent statement pointing to the bigchaindb-js driver would work, would it?

Does  the above explain clearly enough and in sufficient detail what I am trying to do? Which additional information do you need?

Thanks for your patience and your help,

Alex

**Comment by jaimecbernardo** [comment]
Not sure if I understood correctly, but it seems to me this could be feasible by having your main JS file run the driver and setup the communication method for receiving and making the transaction.

Example: Your node script could start the driver and also open a webserver to accept the transaction details from your Java code (In JSON format, for example) and execute the blockchain transaction.

Regarding this question:
> I don't think having a single startNodeWithArguments statement for the creation and execution of a bigchaindb blockchain transaction qith "require" (?) or equivalent statement pointing to the bigchaindb-js driver would work, would it?

Isn't that what this example in the "js-bigchaindb-driver" is doing? https://github.com/bigchaindb/js-bigchaindb-driver/tree/5cb24cff843082f1fd8444453882633dd2f73e3a#example-create-a-transaction

**Comment by adonnini** [comment]
@jaimecbernardo 

Great! Conceptually, what you describe is what I was thinking of trying. However, not knowing much about JS, I wasn't sure how to implement it. Thanks very much. This is very useful. I'll give it a host and let you know.

with regards to the use of the require statement to access the driver, yes that's another way to go about doing what I want. However, it would presume that the driver is loaded or in some way accessible, am I wrong?

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,

When you run `npm install bigchaindb-driver` inside "nodejs-project", it will create a `node_modules` folder with "bigchaindb-driver" inside. When node runs `const driver = require('bigchaindb-driver')`, it will look for "bigchaindb-driver" in "node_modules".

In case I understood what you meant by "the driver is loaded or in some way accessible". The "bigchaindb-driver" module might have some quirks I'm not familiar with and might not be understanding exactly what you mean.

**Comment by adonnini** [comment]
@jaimecbernardo 

Just FYI. As I am writing the code to implement your suggestion, I thought I would try to run the sample app with one startNodeWithArguments with the const driver = require('bigchaindb-driver') statement to start with.

I still had the app copying the bigchaindb-js folder onto the device.

It failed with this error:

Error: Cannot find module 'bigchaindb-driver'
        at Function.Module._resolveFilename (module.js:527:15)
        at Function.Module._load (module.js:476:23)
        at Module.require (module.js:568:17)
        at require (internal/module.js:11:18)
        at [eval]:1:16
        at ContextifyScript.Script.runInThisContext (vm.js:50:33)
        at Object.runInThisContext (vm.js:139:38)
        at Object.<anonymous> ([eval]-wrapper:6:22)
        at Module._compile (module.js:624:30)
        at evalScript (bootstrap_node.js:465:27)

By the way, I re-ran npm install. It completed successfully producing a node_modules folder containing a bigchaindb-drive folder. the path to bigchaindb-driver on the device is:

/data/user/0/com.yourorg.sample/files/bigchaindb-js/node_modules/bigchaindb-driver

When I used device explorer in Android Studio, the path to the bigchaindb-driver folder was:

/data/data/com.yourorg.sample/files/bigchiandb-js/node_modules/bigchaindb-driver

and the folder was fully populated.

What am I missing?

Thanks,

Alex

**Comment by jaimecbernardo** [comment]
Hi @adonnini ,

From your stack trace, it seems that you are trying to run code directly as if you were calling `node -e {some code}`.

The `require` documentation might be helpful: https://nodejs.org/docs/latest-v8.x/api/modules.html#modules_loading_from_node_modules_folders

> If the module identifier passed to require() is not a core module, and does not begin with '/', '../', or './', then Node.js starts at the parent directory of the current module, and adds /node_modules, and attempts to load the module from that location. Node will not append node_modules to a path already ending in node_modules.

So, it would work if you were starting the nodejs runtime from a file, like `/data/user/0/com.yourorg.sample/files/bigchaindb-js/main.js`

For the case where you want to access the "node_modules" without starting the engine a file, you'd have to add the folder where to search the modules to the `NODE_PATH` environment variable before starting the engine, as described in https://nodejs.org/docs/latest-v8.x/api/modules.html#modules_loading_from_the_global_folders

**Comment by adonnini** [comment]
@jaimecbernardo 

Well, success! Using the absolute path worked:

                    startNodeWithArguments(new String[]{"node", "-e",
                            "const driver = require('/data/user/0/com.yourorg.sample/files/bigchaindb-js/node_modules/bigchaindb-driver')\n" +
                                    "const API_PATH = 'https://test.ipdb.io/api/v2/'\n" +
                                    "const alice = new driver.Ed25519Keypair()\n" +
                                    "const tx = driver.Transaction.makeCreateTransaction({ city: 'Berlin, DE', temperature: 22, datetime: new Date().toString() }, { what: 'My first BigchainDB transaction' }, [ driver.Transaction.makeOutput(driver.Transaction.makeEd25519Condition(alice.publicKey))], alice.publicKey)\n" +
                                    "const txSigned = driver.Transaction.signTransaction(tx, alice.privateKey)\n" +
                                    "const conn = new driver.Connection(API_PATH)\n" +
                                    "conn.postTransactionCommit(txSigned)\n" +
                                    "    .then(retrievedTx => console.log('Transaction', retrievedTx.id, 'successfully posted.'))\n"                    });

Now the failure is in the transaction. That's another story and not something you can help me on.

Thank you very much. I will let you know if my declaration of success was premature.

Thanks,

Alex

**Comment by adonnini** [comment]
Hi @jaimecbernardo 

A quick update. I was able to successfully create and execute a bigchaindb blockchain transaction from the sample app. Good progress.

Next, and this might not be possible, in which case I would be up the proverbial creek, is that I would like to pass some data from my Android Java app to the bigchaindb transaction execution script.

Is passing data to the node.js runtime environment from a Java android app possible? If it is, could you point me in the direction of an example/starting point?

Thanks,

Alex Donnini

**Comment by adonnini** [comment]
I guess the question I am asking has two parts:

1) what are the arguments supported by startNodeWithArguments

2) Can a value for a variable used in the script startNodeWithArguments is supposed to run be passed using startNodeWithArguments

FYI, In J2V8 I believe that executeScript and executeJSFunction support passing of content like I would like to be able to. I do realize that your approach may be fundamentally different from the one used to implement J2V8.

Thanks,

Alex Donnini

**Comment by jaimecbernardo** [comment]
Hi @adonnini,
`startNodeWithArguments`  will start node as if it was being started from a Desktop command line with those arguments. (That's why you're able to do `-e` to run a script, which is also the way to do it in Desktop).
So, you could test your script in the Desktop version of node and do the same call to `startNodeWithArguments`.
In terms of passing further information while the engine is running, you'd have to set another way to do it.

We have an old iOS demo where some arguments are passed: https://github.com/janeasystems/node-mobile-react-demo/blob/24eeaef8aabbb1c760384b758c9d5e820118b7d5/ios/PouchAlbum/NodeRunner.mm#L87-L93 from ObjectiveC and read in NodeJS through the minimist module: 
https://github.com/janeasystems/node-mobile-react-demo/blob/24eeaef8aabbb1c760384b758c9d5e820118b7d5/nodeproj/main.js#L222-L225
In this sample, the control file handler argument is a Pipe file handler id which is used to receive information when the App goes to the background and foreground but most of the communication in the sample is done through WebSockets running in an Express Server.

The `nodejs-mobile-react-native` and `nodejs-mobile-cordova` plugins include a native module that is able to create a channel to receive events from Objective-C on iOS and Java on Android, so those might be worth a look.

**Comment by adonnini** [comment]
Thanks @jaimecbernardo 

I found a simpler solution for my problem. Since the transaction payload consists mostly (exclusively) of strings, I defined a transaction creation and execution template. Then I used String.format (Java) to add variable content, and create specific transactions.

So, now I think I am ready to test my app end-to-end, i.e. create a user's mobility profile and location history, and let the user sell them via blockchain.

Thanks,

Alex Donnini

**Comment by DuBistKomisch** [comment]
> > By the way, I guess that setting up multiple threads to run instances of the engine would not make any difference, right?
> 
> Multiple threads would not make any difference, as the engine is shared in the process. Multiple processes should you enable you to do this, however.
> 
> Please let us know of the results you get if you try any of the methods.

Late reply, but I've got this working using a `Service` running in its own process.

I start node in a thread in `onCreate`. If node crashes, the process is killed. If node calls `process.exit` etc the process ends. If it happens to exit gracefully, the next line runs `Process.killProcess(Process.myPid())` to kill the process. So service running is 1:1 with node running. When the service nexts starts, it will be in a new process so node will start successfully.

From another process, a watchdog activity/service/etc can `bindService` with `BIND_AUTO_CREATE`, and observe `onServiceDisconnected` and call `startService` to force the service to restart immediately.

**Comment by jaimecbernardo** [comment]
Hi @DuBistKomisch ,
Thanks for sharing.
That's an interesting solution for this issue in Android :)
